code/tsp_ls2.py

```diff
@@ -2,12 +2,152 @@
 from tsp_types import Trace
 from tsp_types import Node
 from tsp_types import Edge
+from tsp_types import Tracepoint
 import random
 import numpy as np
 import math
 import time
+from multiprocessing import Process, Pipe, Queue
+
+def distance_calculator(nodes):
+    num_locs = len(nodes)
+    mat = np.empty((num_locs,num_locs), dtype=Edge)
+    for x in range(num_locs):
+        for y in range(num_locs):
+            mat[x,y] = Edge(nodes[x],nodes[y])
+            if x == y:
+                mat[x,y].cost = np.inf
+    return mat
+
+def get_min_cost(dest_group,rem_nodes):
+    min_cost = np.inf
+    min_index = np.inf
+    for i in range (0,dest_group.shape[0]):
+        if (dest_group[i].cost < min_cost) and rem_nodes[i] == 1:
+            min_cost = dest_group[i].cost
+            min_index = i
+    return min_cost, min_index
+
+def calculate_route_cost(route,cost_matrix):
+    cost = 0
+    for i in range(0,len(route)-1):
+        cost += cost_matrix[route[i],route[i+1]].cost
+    return cost
+
+def printmat(mat):
+    for i in range(0,len(mat)):
+        for j in range(0,len(mat)):
+            print(mat[i,j].cost, end =" ")  
+        print()   
+
+#This will solve a greedy problem to give us a place to start
+def greedy_heuristic(dist_matrix,index_start):
+    initialdist_matrix = dist_matrix
+    start_index = index_start#random.randint(0,dist_matrix.shape[0]-1)
+    total_cost = 0
+    route = []
+    rem_nodes = np.ones(dist_matrix.shape[0])
+    rem_nodes[start_index] = 0
+    route.append(start_index)
+    while rem_nodes.any(): #while there are still unused locations
+        dist,next_hop = get_min_cost(dist_matrix[start_index,],rem_nodes)
+        route.append(next_hop)
+        start_index = next_hop
+        rem_nodes[next_hop] = 0
+        total_cost += dist
+    total_cost += initialdist_matrix[route[len(route)-1],route[0]].cost
+    route.append(route[0])
+    return route,total_cost    
+
+def accep_criteria(neighbor_route,current_route,cost_matrix,temperature):
+    neighbor_route_cost = calculate_route_cost(neighbor_route,cost_matrix)
+    current_route_cost = calculate_route_cost(current_route,cost_matrix)
```

Comment:
This seems inefficient. Can't you just pass in the cost from the previous iteration?

Reply:
yeah this is bad, fixed now